OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/fieldaccess/reflection/FieldMatcher.scala

```diff
@@ -0,0 +1,150 @@
+/* BSD 2-Clause License - see OPAL/LICENSE for details. */
+package org.opalj
+package tac
+package fpcf
+package analyses
+package fieldaccess
+package reflection
+
+import org.opalj.br.ObjectType
+import org.opalj.br.analyses.ProjectIndexKey
+import org.opalj.br.analyses.SomeProject
+import org.opalj.br.Field
+import org.opalj.br.FieldType
+import org.opalj.tac.fpcf.analyses.cg.V
+import org.opalj.tac.fpcf.analyses.cg.isTypeCompatible
+import org.opalj.value.IsReferenceValue
+
+import scala.collection.immutable.ArraySeq
+
+/**
+ * Used to determine whether a certain field should be considered as a target for a reflective access.
+ * These accesses should be resolved by chaining matchers in [[FieldMatching.getPossibleFields]].
+ *
+ * @author Maximilian Rüsch
+ */
+trait FieldMatcher {
+
+    def initialFields(implicit p: SomeProject): Iterator[Field]
+    def contains(field: Field)(implicit p: SomeProject): Boolean
+    def priority: Int
+}
+
+final class NameBasedFieldMatcher(val possibleNames: Set[String]) extends FieldMatcher {
+
+    override def initialFields(implicit p: SomeProject): Iterator[Field] = {
+        possibleNames.iterator.flatMap(p.get(ProjectIndexKey).findFields)
+    }
+    override def contains(f: Field)(implicit p: SomeProject): Boolean = possibleNames.contains(f.name)
+    override def priority: Int = 2
+}
+
+class ClassBasedFieldMatcher(
+        val possibleClasses:          Set[ObjectType],
+        val onlyFieldsExactlyInClass: Boolean
+) extends FieldMatcher {
+
+    // TODO use a ProjectInformationKey or WeakHashMap to cache fields per project
+    // (for the contains check)
+    private[this] def fields(implicit p: SomeProject): Set[Field] = possibleClasses.flatMap { c =>
+        // todo what about "inherited" fields?
```

Comment:
Yes, what about them?

Reply:
A simple way to include inherited fields here would be to use the `DeclaredFields` here and check the "available" fields declared on the class, i.e. the fields declared on the class or its supertypes. However, this could only be done for defined fields, as we would need information on it being private or not (as private fields cannot be inherited).

Also, when the matcher is given a class and all its subtypes as the possible classes, we would compute the available fields many times over and over for the same classes. That is also included in the TODO one above.

Lets talk about this and the other comment shortly in person.

Reply:
As discussed in person, i will open a separate PR where we address inherited fields properly, similar to `project.instanceMethods`.